subsys/pm/device_runtime.c

```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
 						  PM_DEVICE_STATE_SUSPEND,
 						  NULL, NULL);
 		}
-		ret = 0;
 		goto out;
 	}
 
-	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
+
+	if (!dev->pm->enable) {
+		ret = -ENOTSUP;
+		goto out_unlock;
+	}
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
+		dev->pm->usage++;
+		if (dev->pm->usage < 0) {
+			goto out_unlock;
+		}
+	} else {
+		dev->pm->usage--;
+		if (dev->pm->usage > 1) {
+			goto out_unlock;
+		}
+	}
+
 
 	/* Return in case of Async request */
 	if (pm_flags & PM_DEVICE_ASYNC) {
-		ret = 0;
-		goto out;
+		(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+		goto out_unlock;
 	}
 
-	k_mutex_init(&request_mutex);
-	k_mutex_lock(&request_mutex, K_FOREVER);
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
+	(void)k_mutex_unlock(&dev->pm->lock);
+	pm_device_wait(dev, K_FOREVER);
 
 	/*
 	 * dev->pm->state was set in device_pm_callback(). As the device
 	 * may not have been properly changed to the target_state or another
 	 * thread we check it here before returning.
 	 */
-	ret = target_state == atomic_get(&dev->pm->state) ? 0 : -EIO;
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
```

Comment:
there will not be a need for this additional lock, we are locked above, so basically it should end up like this:

```

	while ((k_work_delayable_is_pending(&dev->pm->work)) ||
		(dev->pm->state == PM_DEVICE_STATE_SUSPENDING) ||
		(dev->pm->state == PM_DEVICE_STATE_RESUMING)) {
		ret = k_condvar_wait(&dev->pm->condvar, &dev->pm->lock,
			       K_FOREVER);
		if (ret != 0) {
			break;
		}
	}
	pm_device_runtime_state_set(dev->pm);
	/*
	 * dev->pm->state was set in device_pm_callback(). As the device
	 * may not have been properly changed to the target_state or another
	 * thread we check it here before returning.
	 */
	ret = target_state == dev->pm->state ? 0 : -EIO;

out_unlock:
	(void)k_mutex_unlock(&dev->pm->lock);
```


The callback will be called synchronously, so the above should work.

Reply:
Ok, let be very clear. This **just** works because we are assuming that the device is doing it synchronous. That is a condition that  we are putting on devices

Reply:
if the device does it asynchronous we may return before the callback be called with a transitional state, RESUMING or SUSPENDING.

Reply:
async calls do not get here, they are now being handled by the workqueue. This is something that can be documented, a device driver might also choose not to call the callback at all, which another pitfall.